### PR TITLE
Refactor a little the code in `launcher/net/` files

### PR DESCRIPTION
**Pull request by flowln**
This PR aims to improve the code in PolyMC's net code, by making code easier to read, and reusing some components common to other parts of the code base (for instance, the `NetAction` logic is very similar to the `Task` logic, so we can sort of merge them in a hierarchy). It also tries to do further work towards #133, so that it can someday be merged :smile: 

Bigger changes could also be made (like using a `weak_ptr` in `ByteArraySink` to try to block a possible free-during-use), but those would incur in a LOT of changes all around the code base (i tried and it was not funny :< ), so I thought it was better not to.

### CODE REVIEW

**Comment by txtsd** [comment]
Thanks for addressing #133

I have a problem with liberal use of `auto` though.

Like it's fine in code like `for(auto it = std::begin(v); it != std::end(v); ++it) {...}` but in almost all other cases it reduces readability.

**Comment by flowln** [comment]
> Like it's fine in code like `for(auto it = std::begin(v); it != std::end(v); ++it) {...}` but in almost all other cases it reduces readability.

Hmm I used to think the same before, but after using it for a while, I started liking it, because it makes almost every function name starts on the same place relative to the start of the signature. Since most times reading the method name is more important than reading its return type for understanding what the method does, it helps when there are a lot of methods in a header that you have to look through to find what can you do with that class.

It's really a matter of opinion though, so if you really don't like it, I can change it back :) Problem is that I already started doing this on other code I modified, so some parts of the code base are already like that :<

**Comment by txtsd** [comment]
That's fair. If no one else has a problem with it, it should be fine.

**Comment by txtsd** [comment]
When you're done with code changes you should add license headers.

File: launcher/InstanceImportTask.cpp [hunk 1]
```diff
@@ -14,27 +14,25 @@
  */
```
Comment (Scrumplex) [inline launcher/InstanceImportTask.cpp]: needs updated license header

File: launcher/minecraft/AssetsUtils.cpp [hunk 2]
```diff
@@ -297,7 +297,7 @@ NetAction::Ptr AssetObject::getDownloadAction()
             auto rawHash = QByteArray::fromHex(hash.toLatin1());
             objectDL->addValidator(new Net::ChecksumValidator(QCryptographicHash::Sha1, rawHash));
         }
```
Comment (Scrumplex) [inline launcher/minecraft/AssetsUtils.cpp:319]: ditto

File: launcher/screenshots/ImgurAlbumCreation.cpp [hunk 3]
```diff
@@ -13,12 +13,12 @@
 ImgurAlbumCreation::ImgurAlbumCreation(QList<ScreenShot::Ptr> screenshots) : NetAction(), m_screenshots(screenshots)
 {
     m_url = BuildConfig.IMGUR_BASE_URL + "album.json";
```
Comment (Scrumplex) [inline launcher/screenshots/ImgurAlbumCreation.cpp:50]: ditto

File: launcher/screenshots/ImgurAlbumCreation.h [hunk 4]
```diff
@@ -24,16 +24,14 @@ class ImgurAlbumCreation : public NetAction
 
 protected
 slots:
```
Comment (Scrumplex) [inline launcher/screenshots/ImgurAlbumCreation.h:61]: ditto

File: launcher/screenshots/ImgurUpload.cpp [hunk 5]
```diff
@@ -13,13 +13,13 @@
 ImgurUpload::ImgurUpload(ScreenShot::Ptr shot) : NetAction(), m_shot(shot)
 {
```
Comment (Scrumplex) [inline launcher/screenshots/ImgurUpload.cpp:49]: ditto

File: launcher/screenshots/ImgurUpload.h [hunk 6]
```diff
@@ -21,7 +21,7 @@ protected
 
 public
```
Comment (Scrumplex) [inline launcher/screenshots/ImgurUpload.h:57]: ditto

File: launcher/tasks/Task.cpp [hunk 7]
```diff
@@ -99,7 +99,7 @@ void Task::emitAborted()
     m_state = State::AbortedByUser;
     m_failReason = "Aborted.";
     qDebug() << "Task" << describe() << "aborted.";
```
Comment (Scrumplex) [inline launcher/tasks/Task.cpp:121]: ditto

File: launcher/tasks/Task.h [hunk 8]
```diff
@@ -52,6 +52,8 @@ class Task : public QObject {
```
Comment (Scrumplex) [inline launcher/tasks/Task.h:68]: ditto

File: launcher/translations/TranslationsModel.cpp [hunk 9]
```diff
@@ -667,7 +667,7 @@ void TranslationsModel::downloadTranslation(QString key)
     auto dl = Net::Download::makeCached(QUrl(BuildConfig.TRANSLATIONS_BASE_URL + lang->file_name), entry);
```
Comment (Scrumplex) [inline launcher/translations/TranslationsModel.cpp:702]: ditto

**Review by Scrumplex — APPROVED** [review]
Thanks a lot!

**Comment by txtsd** [comment]
`AutoMoc: /home/txtsd/git/polymc/PolyMC/launcher/tasks/Task.h:89: Warning: Signals cannot be declared virtual`

Is this not a problem?

File: launcher/net/PasteUpload.cpp [hunk 10]
```diff
@@ -1,3 +1,38 @@
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  PolyMC - Minecraft Launcher
+ *  Copyright (c) 2022 flowln <flowlnlnln@gmail.com>
```
Comment (LennyMcLennington) [inline launcher/net/PasteUpload.cpp]: What copyright do you own here?

File: launcher/net/PasteUpload.h [hunk 11]
```diff
@@ -1,4 +1,40 @@
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  PolyMC - Minecraft Launcher
+ *  Copyright (c) 2022 flowln <flowlnlnln@gmail.com>
```
Comment (LennyMcLennington) [inline launcher/net/PasteUpload.h]: What copyright do you own here?

File: launcher/net/Validator.h [hunk 12]
```diff
@@ -1,3 +1,38 @@
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  PolyMC - Minecraft Launcher
+ *  Copyright (c) 2022 flowln <flowlnlnln@gmail.com>
```
Comment (LennyMcLennington) [inline launcher/net/Validator.h]: What copyright do you own here?

**Review by LennyMcLennington — COMMENTED** [review]
Wouldn't it be better to just put the license header in files that were actually modified by you? I don't see a point in adding a license header to a file not containing any GPL licensed code.

In PasteUpload.cpp and related files it does contain some GPL licensed code from Swirl and I, but the proper header with a more accurate copyright notice is being added in another PR anyway.

**Comment by Scrumplex** [comment]
> `AutoMoc: /home/txtsd/git/polymc/PolyMC/launcher/tasks/Task.h:89: Warning: Signals cannot be declared virtual`
> 
> Is this not a problem?

That's already in develop for a few weeks now.

**Comment by txtsd** [comment]
> > `AutoMoc: /home/txtsd/git/polymc/PolyMC/launcher/tasks/Task.h:89: Warning: Signals cannot be declared virtual`
> > Is this not a problem?
> 
> That's already in develop for a few weeks now.

Might be a good idea to fix it in this PR though.

**Review by Scrumplex — APPROVED** [review]

**Review by txtsd — APPROVED** [review]

**Comment by flowln** [comment]
GitHub built-in editor is horrible and ruined my life. I'll fix this mess later T_T

**Review by timoreo22 — APPROVED** [review]
I still don't like that function declaration style but ok

**Comment by LennyMcLennington** [comment]
My review was marked as resolved despite not being resolved. My mistake for not bringing it up again. Next time please don't add a copyright or license header *at all* unless some of the code is GPL. It doesn't make sense to add a GPL license notice when none of the code is GPL licensed. What my review was saying is please leave the headers alone completely and just let people add the headers when they actually added GPL licensed code.

Now I have to solve unnecessary merge conflicts in my own PR for the license headers because I added license headers where they're actually needed and this one added them where they aren't needed.

And sorry for sounding angry in this comment it's not how I intended it to come out but I don't really know how to edit it to sound less like that, but just know it's not my intention to be rude or anything, I'm just trying to ask that in the future we keep the license headers in files that we actually have GPL code in.

**Comment by txtsd** [comment]
@LennyMcLennington That's my bad. I resolved it after seeing the changes in https://github.com/PolyMC/PolyMC/pull/500/commits/96deb5b09d6729f4b5f3a5c1880b526ee9882326.

**Pull request merged**